subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy

```diff
@@ -174,6 +174,47 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         outputDirectory.list() == [input.name]
     }
 
+    @Issue('https://github.com/gradle/gradle/issues/9904')
+    def "task with broken symlink in inputs are valid"() {
+        def brokenInputDirectory = file('brokenInputDirectory').createLink("brokenInputDirectoryTarget")
+        def inputDirectory = file('inputDirectory').createDir()
+        def brokenInputFile = inputDirectory.file('brokenInputFile').createLink("brokenInputFileTarget")
+        def output = file("output.txt").createFile()
+
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @InputDirectory File brokenInputDirectory
+                @InputDirectory File inputDirectory
+                @InputFile File brokenInputFile
+                @OutputFile File output
+
+                @TaskAction execute() {
+                    output.text = "\${brokenInputDirectory.name} \${brokenInputFile.name} \${inputDirectory.list()}"
+                }
+            }
+            task inputBrokenLinkNameCollector(type: CustomTask) {
+                brokenInputDirectory = file "${brokenInputDirectory}"
+                inputDirectory = file "${inputDirectory}"
+                brokenInputFile = file "${brokenInputFile}"
+                output = file "${output}"
+            }
+        """
+        assert !brokenInputDirectory.exists() && !brokenInputFile.exists() && inputDirectory.exists() && output.text.empty
+
+        when:
+        run 'inputBrokenLinkNameCollector', '--info'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "brokenInputDirectory brokenInputFile [brokenInputFile]"
+        outputContains "Directory '${brokenInputDirectory}' specified for property 'brokenInputDirectory' is a broken symbolic link."
+        outputContains "File '${brokenInputFile}' specified for property 'brokenInputFile' is a broken symbolic link."
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
```

Comment:
Test for changes to the brokenness of the symlinks are missing: So if a symlink is unbroken, then the task needs to be out of date.

Reply:
Wanted to handle incremental stuff in a different PT, but can do it here as well